Support AF_VSOCK in sd_notify and pick up notify_socket from creds

bluca opened the pull request:
If a 'notify_socket' credential is provided, send READY=1 when we finish booting.

This is intended to be used with VSOCK, to notify the hypervisor/VMM, eg on the host:

```
qemu <...> -smbios type=11,value=io.systemd.credential:notify_socket=vsock:2:1234 -device vhost-vsock-pci,id=vhost-vsock-pci0,guest-cid=42
```

(vsock:2:1234 -> send to host on vsock port 1234, default is to send to 0 which is
the hypervisor itself)

Also on the host:

```
$ socat - VSOCK-LISTEN:1234,socktype=5
READY=1
```

poettering commented:
honestly I think we should extend the official spec for $NOTIFY_SOCKET to support vsock natively, and then implement that in sd_notify() natively, for everyone. And document the protocol in the man page.

(i'd probably not support IP, since it's not reliable, IP stack might drop packets, and people really shouldn't have to deal with that. AF_UNIX and AF_VSOCK are both realiable, hence should be OK as transport for sd_notify().

poettering commented on src/core/manager.c:
```diff
@@ -3522,6 +3551,12 @@ static void manager_notify_finished(Manager *m) {
                                                FORMAT_TIMESPAN(total_usec, USEC_PER_MSEC)));
                 }
+
+                /* Send SD_READY=1 to hypervisor. Note that we don't check that we are running in a VM, nor that the passed
```
SD_READY → READY

bluca commented:
> honestly I think we should extend the official spec for $NOTIFY_SOCKET to support vsock natively, and then implement that in sd_notify() natively, for everyone. And document the protocol in the man page.
> 
> (i'd probably not support IP, since it's not reliable, IP stack might drop packets, and people really shouldn't have to deal with that. AF_UNIX and AF_VSOCK are both realiable, hence should be OK as transport for sd_notify().

We cannot send scm rights over vsock (I'd think?) So does it make sense to allow anything to send to vsock? You can't tell who's sending, no?

poettering commented on src/core/manager.c:
```diff
@@ -3522,6 +3551,12 @@ static void manager_notify_finished(Manager *m) {
+                 * notify_socket credential is a VSOCK address. While that's the documented use case, it doesn't hurt to let
+                 * people try weirder things if they want to, after all setting the credential requires privileges over pid1
+                 * so there's no reason not to. */
+                send_notify_ready_via_creds();
```
pid1 already issues sd_notify() at the right places, see manager_print_jobs_in_progress(), manager_send_ready() and thelike. I think we should just make that work. instead of adding another set of similar invocations

poettering commented on src/core/manager.c:
```diff
@@ -3463,6 +3464,34 @@ static void log_taint_string(Manager *m) {
+
+        r = socket_address_pid_notify(&a, /* pid= */ 0, "READY=1", /* fds= */ NULL, /* n_fds= */ 0);
+        if (r < 0)
+                log_warning_errno(r, "Failed to send SD_READY=1 to %s: %m", address);
```
SD_READY → READY

poettering commented on src/core/manager.c:
```diff
@@ -3463,6 +3464,34 @@ static void log_taint_string(Manager *m) {
+        SocketAddress a;
+        int r;
+
+        cred = path_join(SYSTEM_CREDENTIALS_DIRECTORY, "notify_socket");
```
hmm, instead of going via hardcoded path here, i think we should just rely on the fact that the code in src/core/import-creds.c actually sets CREDENTIALS_DIRECTORY env var after it is done. Which means you can just use read_credential() directly, it will to the right thing.

poettering replied:
That said, I wonder if we shouldn't extend import_credentials() to basically do something like this once it is done:

```c
read_credential("notify.socket", &buf, &buf_size);
setenv("NOTIFY_SOCKET", buf, 1);
```

(of course, with more error checking, and with checking that "buf" does not contain a NUL byte)

i.e. that way we first import all the smbios/fw-cfg/… bits and pieces into credentials and then we'd import the notify socket address from there into the NOTIFY_SOCKET env var.

Benefit: we actually just implement the same NOTIFY_SOCKET proto everywhere the same way, just the address gets propagated in differently at different places.

poettering commented:
> We cannot send scm rights over vsock (I'd think?) So does it make sense to allow anything to send to vsock? You can't tell who's sending, no?

well, sd_notify() protocol knows two sides: the receiving service/container/VM manager and the sending service. It's up to the former to figure out whether to trust the latter. The latter doesn't have to care. If $NOTIFY_SOCKET is set to something it should just send to that address, regardless what it is.

The question is how the received decides to trust the sender. For AF_UNIX trust can be established via SCM_CREDENTIALS. I think on AF_VSOCK the typical way to do that would be via VSOCK port nr: a port nr < 1024 indicates a privileged component of the VM, otherwise it could be either.

Which hence suggests: maybe sd_notify() should try to bind to an AF_VSOCK low port before sending its message on it. And handle gracefully (i.e. rely on auto-binding) if that doesn't work. That way we can clearly communicate to the VM manager whether the sending component can be trusted or not.

poettering commented:
> We cannot send scm rights over vsock (I'd think?) So does it make sense to allow anything to send to vsock? You can't tell who's sending, no?

well, sd_notify() protocol knows two sides: the receiving service/container/VM manager and the sending service. It's up to the former to figure out whether to trust the latter. The latter doesn't have to care. If $NOTIFY_SOCKET is set to something it should just send to that address, regardless what it is.

The question is how the received decides to trust the sender. For AF_UNIX trust can be established via SCM_CREDENTIALS. I think on AF_VSOCK the typical way to do that would be via VSOCK port nr: a port nr < 1024 indicates a privileged component of the VM, otherwise it could be either.

Which hence suggests: maybe sd_notify() should try to bind to an AF_VSOCK low port before sending its message on it. And handle gracefully (i.e. rely on auto-binding) if that doesn't work. That way we can clearly communicate to the VM manager whether the sending component can be trusted or not.

bluca commented:
> Which hence suggests: maybe sd_notify() should try to bind to an AF_VSOCK low port before sending its message on it. And handle gracefully (i.e. rely on auto-binding) if that doesn't work. That way we can clearly communicate to the VM manager whether the sending component can be trusted or not.

I'm not sure that would be useful. The trust boundary here is host(hypervisor)/guest, _how_ something is done inside of a VM shouldn't change the threat model at all from the host/hypervisor's point of view.

poettering commented:
> I'm not sure that would be useful. The trust boundary here is host(hypervisor)/guest, _how_ something is done inside of a VM shouldn't change the threat model at all from the host/hypervisor's point of view.

From the host's view it would not. But from the guests's view it would. The guest system layer should have a way to ensure that the ready notification can only come from the system layer, not any apps that run on top of it. AFAIU any process can create/bind/connect AF_VSOCK sockets, regardless of privilege. That would mean that any process can signal ready notifications too – except if we take benefit of the fact that only privileged processes can use the low ports. because if the hypervisor ignores all ready notifications from high ports and the system level in the guest can rely on that we found an effective way so that only the system level of the guess can send the ready notifications.

Or in other words: the trust boundary we'd do the low port binding for is a trust boundary the guest cares for, noone else. But the host can help with that, and enforce a simply restriction that the guest then can use to enforce the trust boundary it actually cares for.

bluca commented:
I've reworked to enhance the generic sd_notify.

> > I'm not sure that would be useful. The trust boundary here is host(hypervisor)/guest, _how_ something is done inside of a VM shouldn't change the threat model at all from the host/hypervisor's point of view.
> 
> From the host's view it would not. But from the guests's view it would. The guest system layer should have a way to ensure that the ready notification can only come from the system layer, not any apps that run on top of it. AFAIU any process can create/bind/connect AF_VSOCK sockets, regardless of privilege. That would mean that any process can signal ready notifications too – except if we take benefit of the fact that only privileged processes can use the low ports. because if the hypervisor ignores all ready notifications from high ports and the system level in the guest can rely on that we found an effective way so that only the system level of the guess can send the ready notifications.
> 
> Or in other words: the trust boundary we'd do the low port binding for is a trust boundary the guest cares for, noone else. But the host can help with that, and enforce a simply restriction that the guest then can use to enforce the trust boundary it actually cares for.

When you setup a network security policy, you don't make assumptions on trust based on the client port, because you don't know what the remote end is doing or running. This is the same. The hypervisor cannot ignore/accept a notification based on whether the port is high or low, as it wouldn't make any sense. Either it chooses to trust the notifications coming from a given guest, or it doesn't (in absence of an authentication mechanism of course, which I'm pretty sure we don't want to add).

bluca commented:
Should I namespace the cred? It's just `notify_socket` now to match `$NOTIFY_SOCKET`, but other creds are namespaces. Maybe `systemd.notify_socket`?

poettering commented on src/basic/socket-util.c:
```diff
@@ -1472,3 +1472,64 @@ int connect_unix_path(int fd, int dir_fd, const char *path) {
+                return -EPROTO;
+
+        return 0;
+}
```
maybe split this in two: 

socket_address_parse_unix() + socket_address_parse_vsock()

I mean, the two transports have so little to do with each other, and the moniker "local" is a bit of a stretch for AF_VSOCK. And confusing too, given that AF_LOCAL exists and just means AF_UNIX...

poettering replied:
The callers would then call this like this:

```c
r = socket_address_parse_unix()
if (r == -EPROTO)
        r = socket_address_parse_vsock()
…
```

which is only a tiny bit longer than the status quo ante, but very clear

poettering commented on src/shared/socket-netlink.c:
```diff
@@ -17,120 +17,74 @@
+        if (r < 0 && r != -EPROTO)
+                return r;
+        if (r >= 0)
+                return 0;
```
I figure you can merge this check into:

```c
if (r != -EPROTO)
        return r;
```

poettering commented on man/sd_notify.xml:
```diff
@@ -368,13 +368,23 @@
+    notification when a virtual machine has finished booting. Note that in
+    case the hypervisor does not support <constant>SOCK_DGRAM</constant>
+    over <constant>AF_VSOCK</constant>, <constant>SOCK_SEQPACKET</constant> will be
+    used instead. The address should be in the form:
```
i'd not say anything about "support". Just say SOCK_DGRAM should be tried first, and SOCK_SEQPACKET as fallback.

bluca replied:
It was non-trivial to find this out, so I want to be explicit as it's a non-obvious low-level details that will trip consumers

bluca replied:
(the important detail here is that it depends on the hypervisor, not on the consumer or the guest)

poettering commented on src/libsystemd/sd-daemon/sd-daemon.c:
```diff
@@ -467,13 +467,34 @@ _public_ int sd_pid_notify_with_fds(
+
+        /* If we didn't get an address, talk to the hypervisor by default, as that's the purpose of this functionality. */
+        if (address.sockaddr.vm.svm_family == AF_VSOCK && address.sockaddr.vm.svm_cid == VMADDR_CID_ANY)
+                address.sockaddr.vm.svm_cid = VMADDR_CID_HYPERVISOR;
```
hmm, shouldn't this be VMADDR_CID_HOST?

(no idea what the precise difference is supposed to be?)

poettering replied:
hmm, i wonder how much sense this automatic fallback makes? maybe just return error here?

bluca replied:
_HYPVERVISOR allows to talk only the hypervisor process, while _HOST allows talking to any process that will listen on the host (I don't actually know how vhost drivers implement this, going by the definition alone). Defaulting to _HOST would mean anything could start listening to state changes in the guests, and might not be desirable (unless it was explicitly configured). I figure talking to the hypervisor is always safe though, as that has privileges over guests by definition. The CID is most often documented as optional with the "right thing" being done if omitted, so this seems to me to be the path of least surprise for consumers.

poettering replied:
it's not clear to me that the "right thing" is to talk to the hypervisor. In my understanding in a libvirt/qemu world, it would be libvrit that gets the messages, not qemu. And i'd assume that _HYPERVISOR ends up only in qemu, not in libvirt. But not sure how this is supposed to work and if i grokked this properly.

But something tells me we should maybe not fall back to a default here if things aren't so clear cut.

bluca replied:
Sure but libvirt also configures the VM, so it can set _HOST. The point is that many documents talking about vsock say the CID is optional, so for consistency we should do the same. And _HOST does not seem like a safe default.

poettering replied:
dunno, still not convinced that doing _HYPERVISOR is really that great a cid to default to here..

I mean, I can see that _HOST is not a safe default. but _HYPERVISOR might be "safe" but I think in almost all cases *wrong*...

so I'd really just fail parsing here.

(but also, i don't care too much. if you insist)

bluca replied:
I insist :-) We can always change it later when the first consumer shows up, it's all academic right now.

poettering replied:
nah, you can't change it later that would be a compat break.

you can't just make something we accept as valid now meaning one thing to mean something totally different.

if we'd generate an error here now, then we'd be on the safe side otoh: we have all options to make it mean whatever we want later.

poettering commented on src/libsystemd/sd-daemon/sd-daemon.c:
```diff
@@ -467,13 +467,34 @@ _public_ int sd_pid_notify_with_fds(
+        /* At the time of writing QEMU does not yet support AF_VSOCK + SOCK_DGRAM and returns
+         * ENODEV. Fallback to SOCK_SEQPACKET in that case. */
+        fd = socket(address.sockaddr.sa.sa_family, SOCK_DGRAM|SOCK_CLOEXEC, 0);
+        if (fd < 0 && errno == ENODEV && address.sockaddr.sa.sa_family == AF_VSOCK) {
```
ENODEV is such a weird error. Maybe add ERRNO_IS_NOT_SUPPORTED() here? (simple to cover for seccomp filters at the very least)

poettering commented on src/libsystemd/sd-daemon/sd-daemon.c:
```diff
@@ -467,13 +467,34 @@ _public_ int sd_pid_notify_with_fds(
+                if (connect(fd, &address.sockaddr.sa, address.size) < 0) {
+                        r = -errno;
+                        goto finish;
+                }
```
hmm, given that the receiver on the host and the hypervisor are two different things and might support different functionality, shouldn't we try SOCK_SEQPACKET also if we can allocate SOCK_DGRAM but not connect() to it? i.e. let's say one day qemu learns SOCK_DGRAM, but libvirt still only does SOCK_SEQPACKET, then you'd not be able to communicate with libvirt's receiver anymore...

bluca replied:
I'm not sure I follow here, there's no connect in the dgram case? The problem is the vsock driver implementation in the hypervisor, that's what's missing support for DGRAM, and looking at the attempts to add it, it doesn't look like a trivial feature to add.
On the other hand, for a consumer simply using AF_VSOCK it's trivial to switch from one protocol to the other.

I figure a consumer should know what the hypervisor is going to support (eg: vmware supports DGRAM as far as I can read), so it can make the right choice.

poettering replied:
again, in a libvirt/qemu world I'd assume that libvirt listens on the AF_VSOCK thing, and qemu does not. But that means qemu and libvirt must agree on the same socket type, because otherwise qemu might open something up that libvirt isn't listening on.

i mean, you are right, forget my comment about connect(), it's really sendmsg() i guess... i.e. if sendmsg() fails on AF_VSOCK/SOCK_DGRAM even though we managed to allocate a socket for that, maybe retry immediately with AF_VSOCK/SOCK_SEQPACKET? because in that case libvirt (as the listener) and qemu (as the transport) can enable support for SOCK_DGRAM independently of each other and things will still work correctly. because otherwise, if qemu starts supporting SOCK_DGRAM but libvirt hasn't been updated yet we'll send messages onto a socket that can't access it.

poettering replied:
i am not sure what "is trivial" has to do with anything. I am just saying that qemu/libvirt might not be updated in lockstep, and hence a robust fallback would make sense.

libvirt implementing both mechanisms right now is problematic, since they can't test it, since they only support qemu and qemu doesn't support SOCK_DGRAM.

This is actually similar for us: we are adding support for SOCK_DGRAM without actually being able to test this. But I think we have a stronger incentive to risk it, since we want compat with vmware and everything else that supports SOCK_DGRAM.

Hence, I am pretty sure that qemu will start supporting SOCK_DGRAM before libvirt will start making use of it, hence I think we should think about the compat strategy here...

bluca replied:
Mh? libvirt supports a gazillion hypervisors, not just qemu, including vmware (and xen, and virtualbox, and hyperv, etc): https://en.wikipedia.org/wiki/Libvirt#Supported_Hypervisors
Hence why if it ever consumes this, it has to do it with both protocols from day one, there's really no other option due to the dependency on the hypervisor + guest kernel.

poettering commented on src/libsystemd/sd-daemon/sd-daemon.c:
```diff
@@ -467,13 +467,34 @@ _public_ int sd_pid_notify_with_fds(
-        if (fd < 0) {
+                msghdr.msg_name = NULL;
+                msghdr.msg_namelen = 0;
+        } else if (fd < 0) {
```
maybe:

```c
if (r < 0) {
        if (!(errno == ENODEV || ERRNO_IS_NOT_SUPPORTED(errno)) || address.sockaddr.sa.sa_family != AF_VSOCK)
               return -errno;
   
        …
```

bluca replied:
Not sure I follow, in this branch the fd is invalid so we always need to error out

poettering replied:
I am just saying that you check for `fd < 0` twice. and I'd nest that: check for it once, and then check if we are in the fallback condition, and if not exit.

bluca replied:
ah ok, got it

poettering commented on src/core/import-creds.c:
```diff
@@ -713,5 +713,15 @@ int import_credentials(void) {
+                _cleanup_free_ char *address = NULL;
+
+                r = read_credential("notify_socket", (void **)&address, /* ret_size= */ NULL);
+                if (r < 0 && r != -ENOENT)
```
also check for ENXIO (i.e. ENOENT: creds passed but the requested one is missing. ENXIO: env var not set, hence no creds passed)

poettering commented on src/core/import-creds.c:
```diff
@@ -713,5 +713,15 @@ int import_credentials(void) {
+                if (r < 0 && r != -ENOENT)
+                        log_warning_errno(r, "Failed to read 'notify_socket' credential, ignoring: %m");
+                else if (r >= 0 && !isempty(address))
+                        setenv("NOTIFY_SOCKET", address, /* replace= */ 1);
```
either cast to (void) or handle error

poettering commented:
> When you setup a network security policy, you don't make assumptions on trust based on the client port, because you don't know what the remote end is doing or running. This is the same. The hypervisor cannot ignore/accept a notification based on whether the port is high or low, as it wouldn't make any sense. Either it chooses to trust the notifications coming from a given guest, or it doesn't (in absence of an authentication mechanism of course, which I'm pretty sure we don't want to add).

It's not about the host "trusting" the guest because of the source port. Ignore that pov. Instead: focus on the guest pov: guest consists roughly of two layers: "privileged system" and "unprivileged user" stuff. The guest is interested that only "privileged system" can send messages to the vmm, but "unprivileged user" stuff cannot. if we establish the rule that the host simply drops all notification messages coming from high ports, then suddenly the guest's privileged layer gets what it wants: it knows that "unprivileged user" stuff cannot just send notifications, because it lacks the privs for sending from low ports.

hence, i still think: we should send these messages from low ports if we can. And that's not about establishing trust on the VMM side, but only about enforcing a requirement to privilege within the guest!

poettering commented:
@berrange btw, this might be interesting to you to maybe support in libvirt? with this PR we'll send ready notifications from PID1 to the VMM if requested, via AF_VSOCK. Which would allow you to run multiple VMs in a series and robustly consume each other services.

(And besides that you'll also get other status updates, such as what systemd is doing right now via STATUS=)

poettering commented on src/basic/socket-util.c:
```diff
@@ -1472,3 +1472,71 @@ int connect_unix_path(int fd, int dir_fd, const char *path) {
         return RET_NERRNO(connect(fd, &sa.sa, salen));
 }
+
+int socket_address_parse_unix(SocketAddress *a, const char *s) {
```
rename `a` to `ret`?

bluca replied:
pre-existing condition in socket_address_parse, but sure

poettering commented on src/basic/socket-util.c:
```diff
@@ -1472,3 +1472,71 @@ int connect_unix_path(int fd, int dir_fd, const char *path) {
+        return 0;
+}
+
+int socket_address_parse_vsock(SocketAddress *a, const char *s) {
```
ditto

poettering commented on src/basic/socket-util.c:
```diff
@@ -1472,3 +1472,71 @@ int connect_unix_path(int fd, int dir_fd, const char *path) {
+
+int socket_address_parse_vsock(SocketAddress *a, const char *s) {
+        /* AF_VSOCK socket in vsock:cid:port notation */
+        const char *cid_start = s + STRLEN("vsock:");
```
i'd really move this down, after the startswith() check below.

static checkers don#t really like pointers that point into invalid memory, and this one might.

poettering replied:
or actually, rewrite the whole thing to be:

```c
cid_start = startswith(s, "vsock:");
if (!cid_start)
        return -EPROTO;
```

i..e get rid of the STRLEN(), taking benefit of the fact that startswith() returns a very useful pointer to the first character after the prefix

bluca commented:
> > When you setup a network security policy, you don't make assumptions on trust based on the client port, because you don't know what the remote end is doing or running. This is the same. The hypervisor cannot ignore/accept a notification based on whether the port is high or low, as it wouldn't make any sense. Either it chooses to trust the notifications coming from a given guest, or it doesn't (in absence of an authentication mechanism of course, which I'm pretty sure we don't want to add).
> 
> It's not about the host "trusting" the guest because of the source port. Ignore that pov. Instead: focus on the guest pov: guest consists roughly of two layers: "privileged system" and "unprivileged user" stuff. The guest is interested that only "privileged system" can send messages to the vmm, but "unprivileged user" stuff cannot. if we establish the rule that the host simply drops all notification messages coming from high ports, then suddenly the guest's privileged layer gets what it wants: it knows that "unprivileged user" stuff cannot just send notifications, because it lacks the privs for sending from low ports.
> 
> hence, i still think: we should send these messages from low ports if we can. And that's not about establishing trust on the VMM side, but only about enforcing a requirement to privilege within the guest!

But there is no enforcement? Nothing stops another process from sending messages from a different port in the guest, binding to a specific port in PID1 has no effect on that. And PID1 is privileged already, so the port it uses doesn't matter, the guest knows this.
In the guest, the way to stop a different process from using AF_VSOCK is via selinux, seccomp and friends - and if those are used, then again the port is irrelevant, as a privileged process can be stopped too.

poettering commented:
> But there is no enforcement? Nothing stops another process from sending messages from a different port in the guest, binding to a specific port in PID1 has no effect on that.

As per docs arbitrary client code cannot bind to low ports, only code with CAP_NET_BIND_SERVICE can do that. Thus there *is* enforcement: only privileged code (i.e. the "privileged system") can bind to low ports, unpriveleged code cannot. This is itself enforces by privileged code (i.e. the kernel), and that is sufficient.

>  And PID1 is privileged already, so the port it uses doesn't matter, the guest knows this.
> In the guest, the way to stop a different process from using AF_VSOCK is via selinux, seccomp and friends - and if those are used, then again the port is irrelevant, as a privileged process can be stopped too.

hmm, i don't follow? relying on selinux/seccomp for this is just bad.

i really dont understand where the confusion is?

the guest kernel can basically already attach one bit of information to all outgoing AF_VSOCK messages that possibly indicates whether the source is considered privileged by the privileged part of the guest. and that's the fact whether the source port is < 1024 or not. If it is < 1024 the source must have possessed CAP_NET_BIND_SERVICE. If the source port is >= 1024, then we cannot derive that information, it might have had it, or might have not.

If the VM host then drops all messages from port numbers >= 1024 then this helps the guest and means that unless the sender had CAP_NET_BIND_SERVICE all messages will be ignore by the VM host. Yay!

bluca commented:
> > But there is no enforcement? Nothing stops another process from sending messages from a different port in the guest, binding to a specific port in PID1 has no effect on that.
> 
> As per docs arbitrary client code cannot bind to low ports, only code with CAP_NET_BIND_SERVICE can do that. Thus there _is_ enforcement: only privileged code (i.e. the "privileged system") can bind to low ports, unpriveleged code cannot. This is itself enforces by privileged code (i.e. the kernel), and that is sufficient.
> 
> > And PID1 is privileged already, so the port it uses doesn't matter, the guest knows this.
> > In the guest, the way to stop a different process from using AF_VSOCK is via selinux, seccomp and friends - and if those are used, then again the port is irrelevant, as a privileged process can be stopped too.
> 
> hmm, i don't follow? relying on selinux/seccomp for this is just bad.
> 
> i really dont understand where the confusion is?

The point is that within the guest, relying on low port numbers is legacy behaviour, and it made some sense back then to avoid any process being able to hog well-known service port numbers for themselves. It makes less sense today when we have rich policy based security enforcement, and even less than that for _clients_ sending things out from a random port, like in this case.

> the guest kernel can basically already attach one bit of information to all outgoing AF_VSOCK messages that possibly indicates whether the source is considered privileged by the privileged part of the guest. and that's the fact whether the source port is < 1024 or not. If it is < 1024 the source must have possessed CAP_NET_BIND_SERVICE. If the source port is >= 1024, then we cannot derive that information, it might have had it, or might have not.
> 
> If the VM host then drops all messages from port numbers >= 1024 then this helps the guest and means that unless the sender had CAP_NET_BIND_SERVICE all messages will be ignore by the VM host. Yay!

Once again, the host should do no such thing, because it would just be security theater - completely meaningless for the host, because there's no way to know what policy a guest is employing from a host's POV. It might have random processes running as root, or with setuid, or with caps, or with a custom kernel, or whatever else, the host really cannot derive any useful security property from the client port number in a packet. So either the host has complete "knowledge" of the guest and full trust, but then the port number doesn't matter, or it doesn't trust it, but then again the port number doesn't matter.
If authentication is needed, then it has to be proper authentication with keys and what not.

poettering commented:
> The point is that within the guest, relying on low port numbers is legacy behaviour, and it made some sense back then to avoid any process being able to hog well-known service port numbers for themselves. It makes less sense today when we have rich policy based security enforcement, and even less than that for _clients_ sending things out from a random port, like in this case.

Nah, this is the wrong way round: selinux and stuff is *second-level* security. A system without selinux on should be reasonably secure already. 

Thus, we must build the system that without a MAC things are reasonably safe already, and that means selinux cannot be the only thing we rely on. 

And we never relied on selinux to fix things for us. for anything. here we shouldn't either.

I am sorry, but if we have no reasonably mechanism in place that ensures that unpriv userspace can't send messages to the hypervisor pretending it was priv system layer we shouldn't merge this at all. 

the high-port/low-port thing on *IP* is garbage, since you cannot trust the transport, and neither peer can trust the other. But this is quite a different situation here for VSOCK: transport is reliable and secure. And hypervisor is reliable and secure. guest can trust both transport and hypervisor (not just *can* actually *must*). the only thing  where trust is lacking is if you look from the hypervisor to the guest. But what I am suggesting here is not about that: the hypervisor shouldn't trust the guest at all. all it should help the guest in enforcing what the guest itself wants to trust.

> Once again, the host should do no such thing, because it would just be security theater - completely meaningless for the host, because there's no way to know what policy a guest is employing from a host's POV. 

Again. This is not about the host trusting the guest. This is solely about the guest itself being sufficiently sure that its unpriv userspace cannot generate messages that are not distinguishable from those generated by priv system layer.

> It might have random processes running as root, or with setuid, or with caps, or with a custom kernel, or whatever else, the host really cannot derive any useful security property from the client port number in a packet. 

again, again and again: IT IS NOT ABOUT THE HOST TRUSTING THE GUEST. 

it's only and exclusively about isolating security layers within the guest. that's all: we want that one layer can do stuff the other cannot.

And yes, it's totally fine if privileged code can pretend to be other privileged code, that's how UNIX works after all. WHat isn't OK however is that unprivileged code can just do what only priivleged code should be able to do.

> So either the host has complete "knowledge" of the guest and full trust, but then the port number doesn't matter, or it doesn't trust it, but then again the port number doesn't matter.
> If authentication is needed, then it has to be proper authentication with keys and what not.

IT'S NOT ABOUT THE HOST TRUSTING THE GUEST!

bluca commented:
Could you please write down what the thread model that you have in mind looks like, and how it is addressed by what you are proposing? Because I am truly not able to understand it

poettering commented on src/libsystemd/sd-daemon/sd-daemon.c:
```diff
@@ -433,19 +433,36 @@ _public_ int sd_is_mq(int fd, const char *path) {
+
+        r = socket_address_parse_vsock(&bind_address, "vsock::1023");
+        if (r < 0)
+                return r;
```
hmm, why not just:

```c
union sockaddr_union sa = {
        .vm.svm_family = AF_VSOCK,
        .vm.svm_port = 1023,
}; 

for (;;) {
        r = RET_NERRNO(bind(fd, &sa.sa, sizeof(sa.vm));
        if (r != -EADDRINUSE || sa.vm.svm_port == 0)
                return r;
        sa.vm_svm_port--;
}
```

i.e. you don't need a whole SocketAddress here, the simple sockaddr_union suffices.

poettering commented on src/libsystemd/sd-daemon/sd-daemon.c:
```diff
@@ -467,17 +484,55 @@ _public_ int sd_pid_notify_with_fds(
+                        goto finish;
+                }
+
+                if (address.sockaddr.sa.sa_family == AF_VSOCK) {
```
this whole block will only be entered if know this is AF_VSOCK, hence this second check is redundant.

poettering commented on src/libsystemd/sd-daemon/sd-daemon.c:
```diff
@@ -467,17 +484,55 @@ _public_ int sd_pid_notify_with_fds(
+
+                if (address.sockaddr.sa.sa_family == AF_VSOCK) {
+                        r = vsock_bind_privileged_port(fd);
+                        if (r < 0)
```
maybe eat up the error if this failed due to a permission problem. i.e. if ERRNO_IS_PRIVILEGE holds, then just eat it up and send from a high port

bluca commented:
> Here's an attack scenario: we have a VM running a container orchestrator, and an attacker gets the ability to run an unprivileged container on it (side by side with bazillions of other containers on the same VM). It wants to offline the whole VM it runs on. So with your current PR, it can simply allocate an AF_VSOCK socket, and send "STOPPING=1\nWATCHDOG=trigger" to the hypervisor/vmm. it can just do that, because the AF_VSOCK stuff you do is entirely unprivileged. Under the assumption that the hypervisor/vmm implements either the sd_notify() service lifecycle proto or the watchdog proto this would cause the hypervisor/vmm to assume the VM was going down (and possibly set a timeout for that to forcibly turn it off eventually) or even immediately trigger an emergency watchdog event to shut it down immediately. Now currently hypervisors/vmms do not implement the receiving side for sd_notify() at all, hence neither the lifecycle proto nor the watchdog proto, but sooner or later they might, because why would they only use it for ready notification if they can use the rest of it too?
> 
> Anyway, such an attack scenario sucks: the fact that unpriv, containerized code can tunnel through to the VMM/hypervisor just like that.
> 
> So, how to address this scenario? My proposal: just make the vmm/hypervisor stupidly drop all incoming sd_notify() messages originating from high ports. That's enough. This means only code in the VM that possesses CAP_NET_BIND_SERVICE can send sd_notify() messages the vmm/hypervisor would accept. Since the container (or any other unpriv userspace) _won't_ have that capability, but the VM's pid1 _will_ have it, we are good: now pid1 of the VM can send these messages and they will be accepted, but arbitrary other stuff running in the same VM cannot.
> 
> Implementation is simple: before issuing connect() just invoke bind() in a loop, with 1023 as first try, counting down, until bind() works, as long as the error seen is EADDRINUSE

I've implemented it now - I'm still not sure I see it in the same way, I'm quite sure that (without strong auth) no vmm/hypervisor should take destructive decisions (a readiness notification is a different story) solely based on an unauthenticated message, it's just too risky. You just need someone to ship a setuid or cap'ed binary and all that goes out of the window... but anyway, I've documented what you said here in the manpage, and implemented it as requested

poettering commented on src/libsystemd/sd-daemon/sd-daemon.c:
```diff
@@ -433,19 +433,36 @@ _public_ int sd_is_mq(int fd, const char *path) {
+
+        do
+                r = RET_NERRNO(bind(fd, &sa.sa, sizeof(sa.vm)));
+        while (r == EADDRINUSE && sa.vm.svm_port-- > 0);
```
EADDRINUSE → -EADDRINUSE

poettering commented on man/sd_notify.xml:
```diff
@@ -368,13 +368,26 @@
+    will be used instead. The address should be in the form:
+    <literal>vsock:CID:PORT</literal>. Note that unlike other uses of vsock,
+    the CID is mandatory and cannot be <literal>VMADDR_CID_ANY</literal>.
+    Note that PID1 will send the VSOCK packets from a privileged port, as an
```
maybe say `… privileged port (i.e. < 1024),…`
